Correct cbrt implementation

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78616
* #78615
* #78350
* #80219
* __->__ #80443

Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

lezcano committed:
Correct cbrt implementation

Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -601,7 +601,10 @@ def _not_impl(*args, **kwargs):
 
 
 def _cbrt_aten(a: torch.Tensor):
```
Might as well fix the annotation while we're here: 

```
def _cbrt_aten(a: torch.Tensor) -> torch.Tensor:
```

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -601,7 +601,10 @@ def _not_impl(*args, **kwargs):
+    if a.is_complex():
+        return pow(a, (1 / 3))
+    else:
+        return torch.where(a >= 0, pow(a, (1 / 3)), -pow(-a, 1 / 3))
```
Want to add a comment for this implementation?

mruberry commented:
cbrt is interesting because it's a prim without a reference or corresponding PyTorch operation, and that means it has little (no?) testing. Should we write a short test in test_prims.py that compares prims.cbrt with scipy.special.cbrt?

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

Lezcano commented:
Addressed the review. @mruberry PTAL

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

IvanYashchuk reviewed (approved):
Thank you for fixing this function!

Lezcano commented on test/test_prims.py:
```diff
@@ -106,6 +111,30 @@ def _wrapper(a):
+        shapes = [(), (0,), (1,), (5,)]
+
+        try:
+            # Sets the default dtype to NumPy's default dtype of double
+            cur_default = torch.get_default_dtype()
+            torch.set_default_dtype(torch.double)
```
I just copied the implementation from the refs test for OpInfos really.

peterbell10 commented on torch/_prims/__init__.py:
```diff
@@ -603,8 +603,18 @@ def _not_impl(*args, **kwargs):
+    # which is a complex number.
+    # For more info see the section Note in
+    # https://en.cppreference.com/w/cpp/numeric/math/cbrt
+    return torch.where(a >= 0, pow(a, (1.0 / 3.0)), -pow(-a, 1.0 / 3.0))
```
How about this implementation using 3 kernels instead of 6, 
```suggestion
    return torch.copysign(torch.pow(a.abs(), 1/3), a)
```

Lezcano replied:
Do we have `copysign` in PrimTorch?

Lezcano replied:
you are completely right, I'll correct it

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

lezcano committed:
Update on "Correct cbrt implementation"


Following
https://github.com/pytorch/pytorch/pull/80219#discussion_r907680368

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

IvanYashchuk commented:
@pytorchbot merge -g
